src/CodeGen_Xtensa.cpp

```diff
@@ -2363,6 +2360,117 @@ HALIDE_ALWAYS_INLINE native_vector_f32_x2 halide_xtensa_concat_from_native(const
+    for (int i = 0; i < Lanes; i++) {
+        tmp[i] = ((const BaseType*)base)[offsets[i]];
+    }
+
+    return *((VectorType *)tmp);
+}
+
+template <typename VectorType, typename OffsetType, typename BaseType, int Lanes>
+HALIDE_ALWAYS_INLINE VectorType gather_load_tcm(const void *base, const OffsetType& offset) {
+    return gather_load_default<VectorType, OffsetType, BaseType, Lanes>(base, offset);
+}
+
+template<>
+HALIDE_ALWAYS_INLINE HALIDE_MAYBE_UNUSED native_vector_i8 gather_load_tcm<native_vector_i8, native_vector_i32_x4, int8_t, VECTOR_WIDTH_U8>(const void *base, const native_vector_i32_x4& offset) {
+  auto addresses1 = native_vector_i32_x2(native_vector_i32_x2::from_native_vector, offset.native_vector[0], offset.native_vector[1]);
+  auto output1 = IVP_GATHERDNX8S(
+    IVP_GATHERANX8S(
+      (const int8_t*) base,
+      convert<native_vector_u16, native_vector_i32_x2>(addresses1)
+    )
+  );
+
+  auto addresses2 = native_vector_i32_x2(native_vector_i32_x2::from_native_vector, offset.native_vector[2], offset.native_vector[3]);
+  auto output2 = IVP_GATHERDNX8S(
+    IVP_GATHERANX8S(
+      (const int8_t*) base,
+      convert<native_vector_u16, native_vector_i32_x2>(addresses2)
+    )
+  );
+
+  // NOTE(aelphy): the intrinsic for gathering 8-bit elements extends them to 16-bit, and the conversion back to 8-bit is needed
+  return convert<native_vector_i8, native_vector_i16_x2>(native_vector_i16_x2(native_vector_i16_x2::from_native_vector, output1, output2));
+}
+
+template<>
+HALIDE_ALWAYS_INLINE HALIDE_MAYBE_UNUSED native_vector_u8 gather_load_tcm<native_vector_u8, native_vector_i32_x4, uint8_t, VECTOR_WIDTH_U8>(const void *base, const native_vector_i32_x4& offset) {
+  auto addresses1 = native_vector_i32_x2(native_vector_i32_x2::from_native_vector, offset.native_vector[0], offset.native_vector[1]);
+  auto output1 = IVP_GATHERDNX8U(
+    IVP_GATHERANX8U(
+      (const uint8_t*) base,
+      convert<native_vector_u16, native_vector_i32_x2>(addresses1)
+    )
+  );
+
+  auto addresses2 = native_vector_i32_x2(native_vector_i32_x2::from_native_vector, offset.native_vector[2], offset.native_vector[3]);
+  auto output2 = IVP_GATHERDNX8U(
+    IVP_GATHERANX8U(
+      (const uint8_t*) base,
+      convert<native_vector_u16, native_vector_i32_x2>(addresses2)
+    )
+  );
+
+  // NOTE(aelphy): the intrinsic for gathering 8-bit elements extends them to 16-bit, and the conversion back to 8-bit is needed
+  return convert<native_vector_u8, native_vector_u16_x2>(native_vector_u16_x2(native_vector_u16_x2::from_native_vector, output1, output2));
+}
+
+template<>
+HALIDE_ALWAYS_INLINE HALIDE_MAYBE_UNUSED native_vector_i16 gather_load_tcm<native_vector_i16, native_vector_i32_x2, int16_t, VECTOR_WIDTH_U16>(const void *base, const native_vector_i32_x2& offset) {
+  return IVP_GATHERDNX16(
+    IVP_GATHERANX16(
+      (const int16_t*) base,
+      convert<native_vector_u16, native_vector_i32_x2>(offset) << 1
```

Comment:
Could you please add a comment that this shift is needed because offsets are expected to be in bytes?

Reply:
done